keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -646,6 +675,85 @@ Pod Status in response to user changing the desired resources in Pod Spec.
   Autoscaler on CPU, memory. This enhancement does not change that limitation.
 
+### Lifecycle Nuances
+
+* Terminated containers can be "resized" in that the resize is permitted by the API, and the Kubelet
+  will accept the changes. This makes race conditions where the container terminates around the
+  resize "fail open", and prevents a resize of a terminated container from blocking the resize of a
+  running container (see [Atomic Resizes](#atomic-resizes)).
+* Resizing pods in a graceful shutdown state is permitted.
+
+### Atomic Resizes
+
+A single resize request can change multiple values, including any or all of:
+* Multiple resource types
+* Requests & Limits
+* Multiple containers
+
+These resource requests & limits can have interdependencies that Kubernetes may not be aware of. For
+example, two containers coordinating work may need to be scaled in tandem. It probably doesn't makes
+sense to scale limits independently of requests, and scaling CPU without memory could just waste
+resources. To mitigate these issues and simplify the design, the Kubelet will treat the requests &
+limits for all containers in the spec as a single atomic request, and won't accept any of the
+changes unless all changes can be accepted. If multiple requests mutate the resources spec before
+the Kubelet has accepted any of the changes, it will treat them as a single atomic request.
+
+`AllocatedResources` only accounts for accepted requests, so the Kubelet will need to record
+allocated limits in its internal checkpoint.
+
+Note: If a second infeasible resize is made before the Kubelet allocates the first resize, there can
+be a race condition where the Kubelet may or may not accept the first resize, depending on whether
+it admits the first change before seeing the second. This race condition is accepted as working as
+intended.
+
+### Sidecars
+
+Sidecars, a.k.a. resizeable InitContainers can be resized the same as regular containers. There are
+no special considerations here. Non-restartable InitContainers cannot be resized.
+
+### QOS Class
+
+A pod's QOS class cannot be changed once the pod is started, independent of any resizes.
+
+To clarify the discussion of QOS Class changes, the following terms are defined:
+
+* "Original QOS Class" - The QOS class that was computed based on the original resource requests &
+  limits when the pod was first created.
+* "Suggested QOS Class" - The QOS class that would be computed based on the current resource
+  requests & limits.
+
+With in-place vertical scaling, the _suggested QOS Class_ must be greater than or equal to the
+_original QOS Class_:
+
+* Guaranteed pods: must maintain `requests == limits`
+* Burstable pods: _can_ be resized such that `requests == limits`, but their original QOS
+class will stay burstable. Must retain at least one CPU or memory request or limit.
+* BestEffort pods: can be freely resized, but stay BestEffort.
```

Comment:
Just to clarify: if a best-effort pod adds a request and limit for say CPU, it's also permitted to later remove them entirely? IOW it would count as "lowering QoS", right?

Reply:
I think the best way to think of it is that the QOS Class __never changes__. So in your example, when you add a request and limit to a best effort pod, it's still BestEffort, even though it's the "shape" of a burstable or guaranteed pod. In this case, removing those requests & limits is allowed, since it's still best effort, and you're just changing it back to the shape of a best effort pod.

Reply:
Thanks for clarifying! Makes sense!